ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -80,41 +86,49 @@ public void run() {
             timer.schedule(refreshIngestionResourceValuesTask, 0, REFRESH_INGESTION_RESOURCES_PERIOD);
 
         } catch (Exception e) {
-            log.error(String.format("Error in initializing ResourceManager: %s.", e.getMessage()), e);
+            log.error("Error in initializing ResourceManager.", e);
             throw e;
         }
     }
 
-    public void clean() {
-        ingestionResources.clear();
+    String getIngestionResource(ResourceTypes resourceType) throws Exception {
+        if (!ingestionResources.containsKey(resourceType)) {
+            // When the value is not available, we need to get the tokens from Kusto (refresh):
+            refreshIngestionResources();
+            try{
+                // If the write lock is locked, then the read will wait here.
+                // In other words if the refresh is running yet, then wait until it ends:
+                ingestionResourcesLock.readLock().lock();
+                if (!ingestionResources.containsKey(resourceType)) {
+                    throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
+                }
+            } finally {
+                ingestionResourcesLock.readLock().unlock();
+            }
+        }
+        return ingestionResources.get(resourceType).nextValue();
```

Comment:
What will happen if ingestionResources is replaced at the middle of all of those operations?